File: Top Interview Questions/L236二叉树的最近公共祖先.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}


function lowestCommonAncestor(
    root?: TreeNode | null,
    p?: TreeNode | null,
    q?: TreeNode | null,
): TreeNode | null {
    if (!root) return null;
    const visited = new WeakSet<TreeNode>();

    const parent = new WeakMap<TreeNode, TreeNode | null>();
    dfs(root, parent);

    while (p) {
        visited.add(p);
        p = parent.get(p);
    }
    while (q) {
        if (visited.has(q)) return q;
        q = parent.get(q);
    }
    return null;
}

function dfs(root: TreeNode, parent: WeakMap<TreeNode, TreeNode | null>) {
    if (root.left) {
        parent.set(root.left, root);
        dfs(root.left, parent);
    }
    if (root.right) {
        parent.set(root.right, root);
        dfs(root.right, parent);
    }
}

File: golang-devildyw/middle/L236二叉树的最近公共祖先.go
package main

func main() {}

//func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
//	if root == nil {
//		return nil
//	}
//	// 当前节点等于任意一个子节点值都立即返回 因为在往下就不是他们的公共父节点了 不再往下找
//	if root.Val == p.Val || root.Val == q.Val {
//		return root
//	}
//
//	// 在左右节点分别找找 q p 在不在
//	left := lowestCommonAncestor(root.Left, p, q)
//	right := lowestCommonAncestor(root.Right, p, q)
//	// 如果左右都包含一个子节点 那么一定是p q 在当前节点的左右两边 当前节点就是他们的最近公共祖先
//	if left != nil && right != nil {
//		// 返回当前节点 当前节点就是他们的公共子节点
//		return root
//	}
//
//	// 要么两边都没有那么直接返回nil即可 在左右两颗子树中都没找到q p 说明就不在当前节点为根树下面
//	if left == nil && right == nil {
//		return nil
//	}
//
//	// 反之 某一边有值 说明pq 两个节点在left 或者right里 并且左右子树返回的结果就是他们的公共祖先了
//	if left == nil {
//		return right
//	}
//	return left
//}

func lowestCommonAncestorNew01(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val == p.Val || root.Val == q.Val {
		return root
	}

	left := lowestCommonAncestorNew01(root.Left, p, q)
	right := lowestCommonAncestorNew01(root.Right, p, q)

	if left != nil && right != nil {
		return root
	}

	if left == nil && right == nil {
		return nil
	}

	if left == nil {
		return right
	}
	return left
}
